diff --git a/src/executors/lint/utility/eslint-utils.js b/src/executors/lint/utility/eslint-utils.js
index e5c40146ef3560047c0e3db88c80f8bf5042602a..86388a67f570249c2a23df47662a878ed774968b 100644
--- a/src/executors/lint/utility/eslint-utils.js
+++ b/src/executors/lint/utility/eslint-utils.js
@@ -38,7 +38,8 @@ async function resolveAndInstantiateESLint(eslintConfigPath, options, useFlatCon
          * not be any html files in the project, so keeping it true would break linting every time
          */
         errorOnUnmatchedPattern: false,
-        reportUnusedDisableDirectives: options.reportUnusedDisableDirectives || undefined,
+        // https://github.com/nrwl/nx/pull/21405
+        // reportUnusedDisableDirectives: options.reportUnusedDisableDirectives || undefined,
     };
     if (useFlatConfig) {
         if (typeof options.useEslintrc !== 'undefined') {
