diff --git a/src/executors/release-publish/release-publish.impl.js b/src/executors/release-publish/release-publish.impl.js
index bfdca6db24e01ee98cf7130b5351e39079222121..5a185e6bdd8f457a0bd66d80cdc8a5483ce69e81 100644
--- a/src/executors/release-publish/release-publish.impl.js
+++ b/src/executors/release-publish/release-publish.impl.js
@@ -33,6 +33,9 @@ async function runExecutor(options, context) {
     const packageJsonPath = (0, path_1.join)(packageRoot, 'package.json');
     const packageJson = (0, devkit_1.readJsonFile)(packageJsonPath);
     const packageName = packageJson.name;
+
+    console.log({ isDryRun, isVerbose: context.isVerbose });
+
     /**
      * Whether or not dynamically replacing local dependency protocols (such as "workspace:*") is supported during `nx release publish` is
      * dependent on the package manager the user is using.
@@ -214,6 +217,12 @@ Please update the local dependency on "${depName}" to be a valid semantic versio
     if (isDryRun) {
         publishCommandSegments.push(`--dry-run`);
     }
+    if (context.isVerbose) {
+        console.log('');
+        console.log('Running publish command:');
+        console.log(publishCommandSegments.join(' '));
+        console.log('');
+    }
     try {
         const output = (0, child_process_1.execSync)(publishCommandSegments.join(' '), {
             maxBuffer: LARGE_BUFFER,
