diff --git a/lib/ajv.d.ts b/lib/ajv.d.ts
index 078364d8ce82f0e0a12b3cf6a72f148f6e26f66e..21cf7bfdcae27b4f343fdc4c18e46e5eee238ac5 100644
--- a/lib/ajv.d.ts
+++ b/lib/ajv.d.ts
@@ -153,7 +153,7 @@ declare namespace ajv {
       parentData?: object | Array<any>,
       parentDataProperty?: string | number,
       rootData?: object | Array<any>
-    ): boolean | PromiseLike<any>;
+    ): boolean;
     schema?: object | boolean;
     errors?: null | Array<ErrorObject>;
     refs?: object;
