diff --git a/lib/rules/no-property-in-node.js b/lib/rules/no-property-in-node.js
index f3e2e0c829685213c7807b34158237f59142173c..11e54e8f6d69e0f06b27bb83ed96a96690b5ed9a 100644
--- a/lib/rules/no-property-in-node.js
+++ b/lib/rules/no-property-in-node.js
@@ -3,6 +3,7 @@
 const typedNodeSourceFileTesters = [
   /@types[/\\]estree[/\\]index\.d\.ts/,
   /@typescript-eslint[/\\]types[/\\]dist[/\\]generated[/\\]ast-spec\.d\.ts/,
+  /packages[/\\]types[/\\]dist[/\\]generated[/\\]ast-spec\.d\.ts/,
 ];
 
 /**
