diff --git a/src/command-line/release/publish.js b/src/command-line/release/publish.js
index c20f0a7ddda51a283ffb5cb8591784cbe07b9a5a..36ff6d2b7312147529538d2204e7e22b31bbdb4f 100644
--- a/src/command-line/release/publish.js
+++ b/src/command-line/release/publish.js
@@ -130,6 +130,10 @@ async function runPublishOnProjects(args, projectGraph, nxJson, projectNames, ex
          */
         process.env.NX_DRY_RUN = 'true';
     }
+    if (args.verbose) {
+        // Ensure the executor context is set appropriately for verbose logging
+        process.env.NX_VERBOSE_LOGGING = 'true';
+    }
     if (args.firstRelease) {
         overrides.firstRelease = args.firstRelease;
     }
