diff --git a/lib/typescript.d.ts b/lib/typescript.d.ts
index 5ee1d5258cb019bddb259c7d31b1ae0156a98f0c..fc5f4d2a757a9fef92b2135148c22bb0cc81b2c1 100644
--- a/lib/typescript.d.ts
+++ b/lib/typescript.d.ts
@@ -4491,8 +4491,8 @@ declare namespace ts {
         JSDocFunctionType = 324,
         JSDocVariadicType = 325,
         JSDocNamepathType = 326,
+        /** @deprecated This was only added in 4.7 */
         JSDoc = 327,
-        /** @deprecated Use SyntaxKind.JSDoc */
         JSDocComment = 327,
         JSDocText = 328,
         JSDocTypeLiteral = 329,
@@ -5066,6 +5066,8 @@ declare namespace ts {
         readonly name: PropertyName;
         readonly questionToken?: QuestionToken;
         readonly type?: TypeNode;
+        /** @deprecated removed in 5.0 but we want to keep it for backwards compatibility checks! */
+        readonly initializer?: Expression | undefined;
     }
     interface PropertyDeclaration extends ClassElement, JSDocContainer {
         readonly kind: SyntaxKind.PropertyDeclaration;
@@ -5091,6 +5093,10 @@ declare namespace ts {
         readonly parent: ObjectLiteralExpression;
         readonly name: PropertyName;
         readonly initializer: Expression;
+        /** @deprecated removed in 5.0 but we want to keep it for backwards compatibility checks! */
+        readonly questionToken?: QuestionToken | undefined;
+        /** @deprecated removed in 5.0 but we want to keep it for backwards compatibility checks! */
+        readonly exclamationToken?: ExclamationToken | undefined;
     }
     interface ShorthandPropertyAssignment extends ObjectLiteralElement, JSDocContainer {
         readonly kind: SyntaxKind.ShorthandPropertyAssignment;
@@ -5098,6 +5104,12 @@ declare namespace ts {
         readonly name: Identifier;
         readonly equalsToken?: EqualsToken;
         readonly objectAssignmentInitializer?: Expression;
+        /** @deprecated removed in 5.0 but we want to keep it for backwards compatibility checks! */
+        readonly modifiers?: NodeArray<Modifier> | undefined;
+        /** @deprecated removed in 5.0 but we want to keep it for backwards compatibility checks! */
+        readonly questionToken?: QuestionToken | undefined;
+        /** @deprecated removed in 5.0 but we want to keep it for backwards compatibility checks! */
+        readonly exclamationToken?: ExclamationToken | undefined;
     }
     interface SpreadAssignment extends ObjectLiteralElement, JSDocContainer {
         readonly kind: SyntaxKind.SpreadAssignment;
@@ -5222,6 +5234,8 @@ declare namespace ts {
     }
     interface FunctionTypeNode extends FunctionOrConstructorTypeNodeBase, LocalsContainer {
         readonly kind: SyntaxKind.FunctionType;
+        /** @deprecated removed in 5.0 but we want to keep it for backwards compatibility checks! */
+        readonly modifiers?: NodeArray<Modifier> | undefined;
     }
     interface ConstructorTypeNode extends FunctionOrConstructorTypeNodeBase, LocalsContainer {
         readonly kind: SyntaxKind.ConstructorType;
@@ -8991,7 +9005,13 @@ declare namespace ts {
     function symbolName(symbol: Symbol): string;
     function getNameOfJSDocTypedef(declaration: JSDocTypedefTag): Identifier | PrivateIdentifier | undefined;
     function getNameOfDeclaration(declaration: Declaration | Expression | undefined): DeclarationName | undefined;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getDecorators from `@typescript-eslint/type-utils`.
+     */
     function getDecorators(node: HasDecorators): readonly Decorator[] | undefined;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getModifiers from `@typescript-eslint/type-utils`.
+     */
     function getModifiers(node: HasModifiers): readonly Modifier[] | undefined;
     /**
      * Gets the JSDoc parameter tags for the node if present.
@@ -9521,7 +9541,13 @@ declare namespace ts {
     function isModuleName(node: Node): node is ModuleName;
     function isBinaryOperatorToken(node: Node): node is BinaryOperatorToken;
     function setTextRange<T extends TextRange>(range: T, location: TextRange | undefined): T;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getModifiers from `@typescript-eslint/type-utils`.
+     */
     function canHaveModifiers(node: Node): node is HasModifiers;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getDecorators from `@typescript-eslint/type-utils`.
+     */
     function canHaveDecorators(node: Node): node is HasDecorators;
     /**
      * Invokes a callback for each child of the given node. The 'cbNode' callback is invoked for all child nodes
