diff --git a/node_modules/@microsoft/api-extractor/lib/api/ExtractorConfig.js b/node_modules/@microsoft/api-extractor/lib/api/ExtractorConfig.js
index a37db0d..841853e 100644
--- a/node_modules/@microsoft/api-extractor/lib/api/ExtractorConfig.js
+++ b/node_modules/@microsoft/api-extractor/lib/api/ExtractorConfig.js
@@ -669,5 +669,5 @@ ExtractorConfig.FILENAME = 'api-extractor.json';
  */
 ExtractorConfig._tsdocBaseFilePath = path.resolve(__dirname, '../../extends/tsdoc-base.json');
 ExtractorConfig._defaultConfig = node_core_library_1.JsonFile.load(path.join(__dirname, '../schemas/api-extractor-defaults.json'));
-ExtractorConfig._declarationFileExtensionRegExp = /\.d\.ts$/i;
+ExtractorConfig._declarationFileExtensionRegExp = /\.d\.(c|m)?ts$/i;
 //# sourceMappingURL=ExtractorConfig.js.map
\ No newline at end of file
