diff --git a/node_modules/@types/esquery/index.d.ts b/node_modules/@types/esquery/index.d.ts
index ba54dba..c101712 100755
--- a/node_modules/@types/esquery/index.d.ts
+++ b/node_modules/@types/esquery/index.d.ts
@@ -4,12 +4,14 @@
 //                 Brad Zacher <https://github.com/bradzacher>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-import { Node } from 'estree';
+import { TSESTree } from '@typescript-eslint/types';
 
 export as namespace esquery;
 
 export = query;
 
+type Node = TSESTree.Node;
+
 /** Query the code AST using the selector string. */
 declare function query(ast: Node, selector: string): Node[];
 
