diff --git a/node_modules/eslint-plugin-import/lib/index.js b/node_modules/eslint-plugin-import/lib/index.js
index 247818e..1b42a54 100644
--- a/node_modules/eslint-plugin-import/lib/index.js
+++ b/node_modules/eslint-plugin-import/lib/index.js
@@ -19,7 +19,8 @@
   'no-named-as-default': require('./rules/no-named-as-default'),
   'no-named-as-default-member': require('./rules/no-named-as-default-member'),
   'no-anonymous-default-export': require('./rules/no-anonymous-default-export'),
-  'no-unused-modules': require('./rules/no-unused-modules'),
+  // temporarily removed for the v8 migration as it crashes the lint run
+  // 'no-unused-modules': require('./rules/no-unused-modules'),

   'no-commonjs': require('./rules/no-commonjs'),
   'no-amd': require('./rules/no-amd'),
