diff --git a/node_modules/eslint-plugin-jest/lib/rules/no-done-callback.js b/node_modules/eslint-plugin-jest/lib/rules/no-done-callback.js
index 8a09e7b..4f7dffa 100644
--- a/node_modules/eslint-plugin-jest/lib/rules/no-done-callback.js
+++ b/node_modules/eslint-plugin-jest/lib/rules/no-done-callback.js
@@ -34,6 +34,7 @@ var _default = (0, _utils.createRule)({
       recommended: 'error',
       suggestion: true
     },
+    hasSuggestions: true,
     messages: {
       noDoneCallback: 'Return a Promise instead of relying on callback parameter',
       suggestWrappingInPromise: 'Wrap in `new Promise({{ callback }} => ...`',
diff --git a/node_modules/eslint-plugin-jest/lib/rules/no-focused-tests.js b/node_modules/eslint-plugin-jest/lib/rules/no-focused-tests.js
index 159e1fb..b167a83 100644
--- a/node_modules/eslint-plugin-jest/lib/rules/no-focused-tests.js
+++ b/node_modules/eslint-plugin-jest/lib/rules/no-focused-tests.js
@@ -36,6 +36,7 @@ var _default = (0, _utils.createRule)({
       recommended: 'error',
       suggestion: true
     },
+    hasSuggestions: true,
     messages: {
       focusedTest: 'Unexpected focused test.',
       suggestRemoveFocus: 'Remove focus from test.'
diff --git a/node_modules/eslint-plugin-jest/lib/rules/prefer-expect-assertions.js b/node_modules/eslint-plugin-jest/lib/rules/prefer-expect-assertions.js
index cf064ac..c47dc4a 100644
--- a/node_modules/eslint-plugin-jest/lib/rules/prefer-expect-assertions.js
+++ b/node_modules/eslint-plugin-jest/lib/rules/prefer-expect-assertions.js
@@ -29,6 +29,7 @@ var _default = (0, _utils.createRule)({
       recommended: false,
       suggestion: true
     },
+    hasSuggestions: true,
     messages: {
       hasAssertionsTakesNoArguments: '`expect.hasAssertions` expects no arguments',
       assertionsRequiresOneArgument: '`expect.assertions` excepts a single argument of type number',
diff --git a/node_modules/eslint-plugin-jest/lib/rules/prefer-strict-equal.js b/node_modules/eslint-plugin-jest/lib/rules/prefer-strict-equal.js
index 95136e6..c1179c6 100644
--- a/node_modules/eslint-plugin-jest/lib/rules/prefer-strict-equal.js
+++ b/node_modules/eslint-plugin-jest/lib/rules/prefer-strict-equal.js
@@ -16,6 +16,7 @@ var _default = (0, _utils.createRule)({
       recommended: false,
       suggestion: true
     },
+    hasSuggestions: true,
     messages: {
       useToStrictEqual: 'Use `toStrictEqual()` instead',
       suggestReplaceWithStrictEqual: 'Replace with `toStrictEqual()`'
