diff --git a/node_modules/eslint-plugin-deprecation/dist/rules/deprecation.js b/node_modules/eslint-plugin-deprecation/dist/rules/deprecation.js
index cbb334d..dcbfa5d 100644
--- a/node_modules/eslint-plugin-deprecation/dist/rules/deprecation.js
+++ b/node_modules/eslint-plugin-deprecation/dist/rules/deprecation.js
@@ -15,7 +15,7 @@ Object.defineProperty(exports, "__esModule", { value: true });
  * along with this program; if not, write to the Free Software Foundation,
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
-const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
+const experimental_utils_1 = require("@typescript-eslint/utils");
 const tsutils_1 = require("tsutils");
 const ts = require("typescript");
 const stringifyJSDocTagInfoText_1 = require("../utils/stringifyJSDocTagInfoText");
