diff --git a/node_modules/react-split-pane/index.d.ts b/node_modules/react-split-pane/index.d.ts
index d116f54..20a132e 100644
--- a/node_modules/react-split-pane/index.d.ts
+++ b/node_modules/react-split-pane/index.d.ts
@@ -6,6 +6,7 @@ export type Split = 'vertical' | 'horizontal';
 
 export type SplitPaneProps = {
   allowResize?: boolean;
+  children?: React.ReactNode;
   className?: string;
   primary?: 'first' | 'second';
   minSize?: Size;
