# Temporary patch. See https://github.com/microsoft/TypeScript/issues/63228
diff --git a/lib/typescript.js b/lib/typescript.js
--- a/lib/typescript.js
+++ b/lib/typescript.js
@@ -127654,6 +127654,9 @@
   (_g = tracing) == null ? void 0 : _g.push(tracing.Phase.Program, "tryReuseStructureFromOldProgram", {});
   structureIsReused = tryReuseStructureFromOldProgram();
   (_h = tracing) == null ? void 0 : _h.pop();
+  if (structureIsReused === 2) {
+    oldProgram.libFiles.forEach((l) => libFiles.add(l))
+  }
   if (structureIsReused !== 2 /* Completely */) {
     processingDefaultLibFiles = [];
     processingOtherFiles = [];
@@ -127790,6 +127793,7 @@
   resolvedModulesProcessing = void 0;
   resolvedTypeReferenceDirectiveNamesProcessing = void 0;
   const program = {
+    libFiles,
     getRootFileNames: () => rootNames,
     getSourceFile,
     getSourceFileByPath,
