diff --git a/node_modules/typescript/lib/typescript.d.ts b/node_modules/typescript/lib/typescript.d.ts
index 374e65a..9d5a88c 100644
--- a/node_modules/typescript/lib/typescript.d.ts
+++ b/node_modules/typescript/lib/typescript.d.ts
@@ -368,8 +368,8 @@ declare namespace ts {
         JSDocFunctionType = 320,
         JSDocVariadicType = 321,
         JSDocNamepathType = 322,
+        /** @deprecated This was only added in 4.7 */
         JSDoc = 323,
-        /** @deprecated Use SyntaxKind.JSDoc */
         JSDocComment = 323,
         JSDocText = 324,
         JSDocTypeLiteral = 325,
@@ -4522,7 +4522,13 @@ declare namespace ts {
     function symbolName(symbol: Symbol): string;
     function getNameOfJSDocTypedef(declaration: JSDocTypedefTag): Identifier | PrivateIdentifier | undefined;
     function getNameOfDeclaration(declaration: Declaration | Expression | undefined): DeclarationName | undefined;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getDecorators from `@typescript-eslint/type-utils`.
+     */
     function getDecorators(node: HasDecorators): readonly Decorator[] | undefined;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getModifiers from `@typescript-eslint/type-utils`.
+     */
     function getModifiers(node: HasModifiers): readonly Modifier[] | undefined;
     /**
      * Gets the JSDoc parameter tags for the node if present.
@@ -5027,7 +5033,13 @@ declare namespace ts {
     function isModuleName(node: Node): node is ModuleName;
     function isBinaryOperatorToken(node: Node): node is BinaryOperatorToken;
     function setTextRange<T extends TextRange>(range: T, location: TextRange | undefined): T;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getModifiers from `@typescript-eslint/type-utils`.
+     */
     function canHaveModifiers(node: Node): node is HasModifiers;
+    /**
+     * @deprecated don't use this directly as it does not exist pre-4.8; instead use getDecorators from `@typescript-eslint/type-utils`.
+     */
     function canHaveDecorators(node: Node): node is HasDecorators;
     /**
      * Invokes a callback for each child of the given node. The 'cbNode' callback is invoked for all child nodes
