---
description: 'Enforce default parameters to be last.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/default-param-last** for documentation.

:::danger Deprecated

This rule has been deprecated because the base [`default-param-last`](https://eslint.org/docs/rules/default-param-last) rule added support for TypeScript syntax.
There is no longer any need to use this extension rule.
It will be removed in the next major version of typescript-eslint.

:::

It adds support for optional parameters.

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
function f(a = 0, b: number) {}
function f(a: number, b = 0, c: number) {}
function f(a: number, b?: number, c: number) {}
class Foo {
  constructor(
    public a = 10,
    private b: number,
  ) {}
}
class Foo {
  constructor(
    public a?: number,
    private b: number,
  ) {}
}
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
function f(a = 0) {}
function f(a: number, b = 0) {}
function f(a: number, b?: number) {}
function f(a: number, b?: number, c = 0) {}
function f(a: number, b = 0, c?: number) {}
class Foo {
  constructor(
    public a,
    private b = 0,
  ) {}
}
class Foo {
  constructor(
    public a,
    private b?: number,
  ) {}
}
```

</TabItem>
</Tabs>
