---
description: 'Disallow using code marked as `@deprecated`.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-deprecated** for documentation.

First description.

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
Incorrect code.
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
Correct code.
```

</TabItem>
</Tabs>

## When Not To Use It

Something about stuff.

## Related To

- [`eslint-plugin-deprecation`](https://github.com/gund/eslint-plugin-deprecation)
