---
description: 'Disallow extraneous union types in return type annotations.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-extraneous-return-types** for documentation.

Explicit function return type annotations can sometimes be helpful in TypeScript to describe a more precise return type than what the type checker would infer.
Functions that can return one of multiple types of values generally use union types to describe their return type.
This lint rule checks for "extraneous", or unnecessary, constituents of union type returns: those that don't describe a type the function can return.

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
function test1(): string | null {
  return 'string';
}

function test2(): Promise<string | null> {
  return Promise.resolve('string');
}

class A {
  test(): string | null {
    return 'string';
  }
}
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
function test1(): string {
  return 'string';
}

function test2(): Promise<string> {
  return Promise.resolve('string');
}

class A {
  test(): string {
    return 'string';
  }
}
```

</TabItem>
</Tabs>

## When Not To Use It

This rule can be completely redundant for projects without explicit return type annotations on functions. It works best when coupled with other rules that enforce explicit return type annotations such as [`explicit-function-return-type`](https://typescript-eslint.io/rules/explicit-function-return-type) or [`explicit-module-boundary-types`](https://typescript-eslint.io/rules/explicit-module-boundary-types).
