---
description: 'Disallow using the spread operator when it might cause unexpected behavior.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-misused-spread** for documentation.

The [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) (`...`) is a powerful feature in JavaScript that can be misused in ways not always detectable by TypeScript. This rule disallows using the spread operator on types where spreading can lead to unexpected behavior.

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
declare const userName: string;
const chars = [...userName];

declare const arr: number[];
const arrSpread = { ...arr };

declare const set: Set<number>;
const setSpread = { ...set };

declare const map: Map<string, number>;
const mapSpread = { ...map };

declare function getObj(): { a: 1; b: 2 };
const getObjSpread = { ...getObj };
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
declare const userName: string;
const chars = userName.split('');

declare const arr: number[];
const arrSpread = [...arr];

declare const set: Set<number>;
const setSpread = [...set];

declare const map: Map<string, number>;
const mapObject = Object.fromEntries(map);

declare function getObj(): { a: 1; b: 2 };
const getObjSpread = { ...getObj() };
```

</TabItem>
</Tabs>

## Options

### `allow`

This option allows marking specific types as "safe" to be spread. It takes an
array of type specifiers to consider safe.

Each item in the array must have one of the following forms:

- A type defined in the current file, or globally available (`"SafeType"`)
- A type defined in a file (`{ from: "file", name: "SafeString", path: "src/safe-string.ts" }` with `path` being an optional path relative to the project root directory)
- A type from the default library (`{ from: "lib", name: "BrandedArray" }`)
- A type from a package (`{ from: "package", name: "ThisIsSafe", package: "safe-lib" }`, this also works for types defined in a typings package).

Examples of a configuration for this option:

```json
"@typescript-eslint/no-misused-spread": [
  "error",
  {
    "allowForKnownSafeIterables": [
      "SafeType",
      { "from": "file", "name": "SafeString", "path": "src/safe-string.ts" },
      { "from": "lib", "name": "BrandedArray" },
      { "from": "package", "name": "ThisIsSafe", "package": "safe-lib" }
    ]
  }
]
```

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
type UnsafeIterable = Iterable<number>;

declare const iterable: UnsafeIterable;

const spreadIterable = { ...iterable };

type UnsafeBrandedString = string & { __brand: 'unsafe' };

declare const brandedString: UnsafeBrandedString;

const spreadBrandedString = { ...brandedString };
```

</TabItem>
<TabItem value="âœ… Correct">

```ts option='{"allow":["SafeIterable", "BrandedString"]}'
type SafeIterable = Iterable<number>;

declare const iterable: SafeIterable;

const spreadIterable = { ...iterable };

type BrandedString = string & { __brand: 'safe' };

declare const brandedString: BrandedString;

const spreadBrandedString = { ...brandedString };
```

</TabItem>
</Tabs>

## When Not To Use It

If you intentionally want to use the spread operator in those cases, and expect
the specific behavior that comes with it, you might not want this rule.
For example, when you want to spread an array into an object and expect the
result to be an object with the array elements as values and the array indices
as keys.
