---
description: 'Disallow using `Object` methods on `Map` or `Set` when it might cause unexpected behavior.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-misused-object-likes** for documentation.

Object methods like `Object.keys` and `Object.values` are intended for plain objects. Using them on `Map` or `Set` is a common mistake that may lead to unexpected behavior.

This rule disallows such usage, encouraging correct alternatives.

<Tabs>
  <TabItem value="âŒ Incorrect">
		```ts
      const mySet = new Set(['foo', 'bar']);
			Object.values(mySet);

      const myMap = new Map([['foo', 'bar'], ['hello', 'world']]);
    		Object.entries(myMap);
    	```
    </TabItem>

  <TabItem value="âœ… Correct">
		```ts
      const mySet = new Set(['foo', 'bar']);
			mySet.keys();

      const myMap = new Map([['foo', 'bar'], ['hello', 'world']]);
    		myMap.entries();
    	```
    </TabItem>

</Tabs>

## When Not To Use It

If your application extends the `Map` or `Set` types in unusual ways or manipulating their class prototype chains, you might not want this rule.

You might consider using [ESLint disable comments](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1) for those specific situations instead of completely disabling this rule.
