---
description: 'Disallow using the spread operator when it might cause unexpected behavior.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-misused-spread** for documentation.

[Spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) (`...`) is a JavaScript feature for creating an object with the joined properties of one or more other objects.
TypeScript allows spreading objects whose properties are not typically meant to be enumerated, such as arrays and class instances.

This rule disallows using the spread syntax on values whose types indicate doing so may cause unexpected behavior.
That includes the following cases:

- Spreading a string into an array.
- Spreading a `Promise` into an object.
  You probably meant to `await` it.
- Spreading a function without properties into an object.
  You probably meant to call it.
- Spreading a class instance into an object.
  This does not faithfully copy the instance because only its own properties are copied, but the inheritance chain is lost, including all its methods.
- Spreading a class (defined with a `class` declaration or `class` expression) into an object.
  This copies all static own properties of the class, but none of the inheritance chain.
- Spreading an iterable (`Map`, `Array`, etc.) into an object.
  Iterable objects usually do not have meaningful enumerable properties and you probably meant to spread it into an array instead.

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
declare const userName: string;
const chars = [...userName];

declare const arr: number[];
const arrSpread = { ...arr };

declare const set: Set<number>;
const setSpread = { ...set };

declare const map: Map<string, number>;
const mapSpread = { ...map };

declare function getObj(): { a: 1; b: 2 };
const getObjSpread = { ...getObj };
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
declare const userName: string;
const chars = userName.split('');

declare const arr: number[];
const arrSpread = [...arr];

declare const set: Set<number>;
const setSpread = [...set];

declare const map: Map<string, number>;
const mapObject = Object.fromEntries(map);

declare function getObj(): { a: 1; b: 2 };
const getObjSpread = { ...getObj() };
```

</TabItem>
</Tabs>

## Options

### `allow`

{/* insert option description */}

This option allows marking specific types as "safe" to be spread. It takes an
array of type specifiers to consider safe.

This option takes the shared [`TypeOrValueSpecifier` format](/packages/type-utils/type-or-value-specifier).

Examples of a configuration for this option:

```json
"@typescript-eslint/no-misused-spread": [
  "error",
  {
    "allowForKnownSafeIterables": [
      "SafeType",
      { "from": "file", "name": "SafeString", "path": "src/safe-string.ts" },
      { "from": "lib", "name": "BrandedArray" },
      { "from": "package", "name": "ThisIsSafe", "package": "safe-lib" }
    ]
  }
]
```

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
type UnsafeIterable = Iterable<number>;

declare const iterable: UnsafeIterable;

const spreadIterable = { ...iterable };

type UnsafeBrandedString = string & { __brand: 'unsafe' };

declare const brandedString: UnsafeBrandedString;

const spreadBrandedString = { ...brandedString };
```

</TabItem>
<TabItem value="âœ… Correct">

```ts option='{"allow":["SafeIterable", "BrandedString"]}'
type SafeIterable = Iterable<number>;

declare const iterable: SafeIterable;

const spreadIterable = { ...iterable };

type BrandedString = string & { __brand: 'safe' };

declare const brandedString: BrandedString;

const spreadBrandedString = { ...brandedString };
```

</TabItem>
</Tabs>

## When Not To Use It

If you intentionally want to use the spread operator in those cases, and expect
the specific behavior that comes with it, you might not want this rule.
For example, when you want to spread an array into an object and expect the
result to be an object with the array elements as values and the array indices
as keys.

If your use cases for unusual spreads only involve a few types, you might consider using
[ESLint disable comments](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1)
and/or the [`allow` option](#allow) instead of completely disabling this rule.

## Further Reading

- [Strings Shouldn't Be Iterable By Default](https://www.xanthir.com/b4wJ1)
