---
description: 'Disallow using confusing built-in primitive class wrappers.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/no-uppercase-wrapper-types** for documentation.

The JavaScript language includes a set of primitives such as `boolean` and `number`s.
JavaScript also contains a "wrapper" class object for each of those primitives, such as [`Boolean`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) and [`Number`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number).
TypeScript represents the primitive in lower-case types and instances of the wrapper classes with upper-case.

Using the primitives like `0` instead of class constructors like `Number(0)` is generally considered a JavaScript best practice.
As a result, using the lower-case primitive type names like `number` instead of class names like `Number` is generally considered a TypeScript best practice.

Similarly, TypeScript's `Function` type allows any function-like value.
It's generally better to specify function parameter and return types with function type syntax.

Examples of code for this rule:

<Tabs>
<TabItem value="❌ Incorrect">

```ts
let myBigInt: BigInt;
let myBoolean: Boolean;
let myNumber: Number;
let myObject: Object;
let myString: String;
let mySymbol: Symbol;

let myFunction: Function;
myFunction = () => {};
```

</TabItem>
<TabItem value="✅ Correct">

```ts
let myBigint: bigint;
let myBoolean: boolean;
let myNumber: number;
let myObject: object;
let myString: string;
let mySymbol: symbol;

let myFunction: () => void;
myFunction = () => {};
```

</TabItem>
</Tabs>

## When Not To Use It

If your project is a rare one that intentionally deals with the class equivalents of primitives, it might not be worthwhile to use this rule.
You might consider using [ESLint disable comments](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1) for those specific situations instead of completely disabling this rule.

## Further Reading

- [MDN documentation on `string` primitives and `String` objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#string_primitives_and_string_objects)

## Related To

- [`ban-types`](./ban-types.mdx)
- [`no-empty-object-type`](./no-empty-object-type.mdx)
- [`no-restricted-types`](./no-restricted-types.mdx)
