---
description: 'Disallow default values that will never be used.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-useless-default-assignment** for documentation.

[Default parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters) and [default values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#default_value) are only used if the parameter or property is `undefined`.
That can happen when a value is missing, or when one is provided and set to `undefined`.
If a non-`undefined` value is guaranteed to be provided, then there is no need to define a default.

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
function Bar({ foo = '' }: { foo: string }) {
  return foo;
}

const { foo = '' } = { foo: 'bar' };

[1, 2, 3].map((a = 42) => a + 1);
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
function Bar({ foo = '' }: { foo?: string }) {
  return foo;
}

const obj: { a?: string } = {};
const { a = 'default' } = obj;

const obj: { a: string | undefined } = { a: undefined };
const { a = 'default' } = obj;

[1, 2, 3, undefined].map((a = 42) => a + 1);
```

</TabItem>
</Tabs>

## When Not To Use It

If your codebase is still onboarding to TypeScript and/or has many existing `any`s or areas of loosely typed code, it may be difficult to enable this rule.
You might consider using ESLint disable comments for those specific situations instead of completely disabling this rule.
