---
description: 'Disallow type assertions that widen a type.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-unsafe-type-assertion** for documentation.

This rule forbids using type assertions to narrow a type, as it can lead to unsafe assumptions that bypass TypeScriptâ€™s type-checking.

Instead, it's better to rely on type guards to ensure type safety and avoid potential runtime errors caused by unsafe type assertions.

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
function f() {
  return Math.random() < 0.5 ? 42 : 'oops';
}

const z = f() as number;

const items = [1, '2', 3, '4'];

const number = items[0] as number;
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
function f() {
  return Math.random() < 0.5 ? 42 : 'oops';
}

const z = f() as number | string | boolean;

const items = [1, '2', 3, '4'];

const number = items[0] as number | string | undefined;
```

</TabItem>
</Tabs>

## When Not To Use It

If your codebase has many unsafe type assertions, then it may be difficult to enable this rule.
It may be easier to skip the `no-unsafe-*` rules pending increasing type safety in unsafe areas of your project.
You might consider using [ESLint disable comments](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1) for those specific situations instead of completely disabling this rule.

This rule requires TypeScript v5.4 or above; you should not use it if your TypeScript version is lower.

## Further Reading

- More on https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions
