---
description: 'Disallow default values that will never be used.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-useless-default-assignment** for documentation.

[Default parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters) and [default values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#default_value) are only used if the parameter or property is `undefined` (either because its value is `undefined` or because it is missing).
If the type of a parameter or property can never be `undefined`, then the default is never actually used.

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
function Bar({ foo = '' }: { foo: string }) {
  return foo;
}

const { foo = '' } = { foo: 'bar' };

[1, 2, 3].map((a = 42) => a + 1);
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
function Bar({ foo = '' }: { foo?: string }) {
  return foo;
}

const obj: { a?: string } = {};
const { a = 'default' } = obj;

const obj: { a: string | undefined } = { a: undefined };
const { a = 'default' } = obj;

[1, 2, 3, undefined].map((a = 42) => a + 1);
```

</TabItem>
</Tabs>

## When Not To Use It

If your project is not accurately typed or `strictNullChecks` is disabled.
