# Requires expressions of type void to appear in statement position (`no-void-expression`)

TODO

## Examples

Examples of **incorrect** code for this rule:

```ts
// TODO
```

Examples of **correct** code for this rule:

```tsx
// TODO
```

## When Not To Use It

TODO

## Related to

- TSLint: ['no-void-expression'](https://palantir.github.io/tslint/rules/no-void-expression/)
