---
description: 'Disallow async functions which do not return promises and have no `await` expression.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/require-await** for documentation.

This rule extends the base [`eslint/require-await`](https://eslint.org/docs/rules/require-await) rule.
It uses type information to allow promise-returning functions to be marked as `async` without containing an `await` expression.

:::note
Unlike the base rule, our implementation also supports [async generator functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function*).
:::

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
async function returnNumber() {
  return 1;
}

async function* asyncGenerator() {
  const num = Promise.resolve(1);
  yield 1 + num;
}

const num = returnNumber();
const callAsyncGenerator = () => asyncGenerator();
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
function returnNumber() {
  return 1;
}

async function* asyncGenerator() {
  const num = await Promise.resolve(1);
  yield 1 + num;
}

const num = () => returnNumber();
const callAsyncGenerator = () => asyncGenerator();
```

</TabItem>
</Tabs>
