---
description: 'Enforce classes are fully assignable to any interfaces they implement.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/strict-interface-implementation** for documentation.

todo explain

## Examples

<Tabs>
<TabItem value="âŒ Incorrect">

```ts
todo;
```

</TabItem>
<TabItem value="âœ… Correct">

```ts
todo;
```

</TabItem>
</Tabs>

## When Not To Use It

todo

## Further Reading

- [TypeScript Handbook > Classes](https://www.typescriptlang.org/docs/handbook/2/classes.html)
