# `your-rule-name`

Your rule description here.

## Rule Details

To fill out: tell us more about this rule.

<!--tabs-->

### ❌ Incorrect

```ts
// To fill out: incorrect code
```

### ✅ Correct

```ts
// To fill out: correct code
```

## Options

This rule is not configurable.

```jsonc
// .eslintrc.json
{
  "rules": {
    "@typescript-eslint/your-rule-name": "error"
  }
}
```

If not configurable: This rule is not configurable.

If configurable...

```ts
type Options = {
  someOption?: boolean;
};

const defaultOptions: Options = {
  someOption: false,
};
```

## When Not To Use It

To fill out: why wouldn't you want to use this rule?
For example if this rule requires a feature released in a certain TS version.
